### PR TITLE
Ports - Steph

### PR DESCRIPTION
**Pull request by smarchante1**
# JS Adagrams
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
--- | ---
What patterns were you able to use from your Ruby knowledge to apply to JavaScript? |  Most of the logic of conditional statements seemed to remain the same. Accessing items from JS objects was very similar to the Ruby hash.
Did you need to use different strategies to find information online about JavaScript than you would for Ruby? |  I needed to search keywords for what I wanted to achieve rather than search ".foo ruby to javascript". I found that a lot of Ruby built in methods are not used in JS and it was more helpful to search what I was trying to do than method names.
What was a challenge you were able to overcome on this assignment? | It was challenging to figure out how to create the player hand. In Ruby adagrams I used .sample which made randomizing letters really simple. For JS i had to use `Math.floor(Math.random() * (letters.length)` to even begin the randomizing logic.
What has been interesting and positive about learning a new programming language? |  I love the fact that I can see how far I've come from February. There are areas of JS that are still quite challenging, but I do feel that I'm picking it up much easier than I did Ruby.
What is something to focus on your learnings in JavaScript in the next week? | I want to continue practicing with syntax ({}, ;, etc..). I think that will continue to be a challenge for me.

### CODE REVIEW

**Comment by tildeee** [comment]
# JS Adagrams
## What We're Looking For



Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | x
**Code Requirements** |
`drawLetters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `drawLetters` pass | x
`usesAvailableLetters` method | x
All tests for `usesAvailableLetters` pass | x
`scoreWord` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `scoreWord` pass | x
`highestScoreFrom` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highestScoreFrom` pass | x
**Overall** |



Well done on this project, Steph!


Overall, your solutions were exactly as I expected (which is a good thing), and your code style was really great! I definitely am adding a lot of comments about variables being `let` instead of `const`, but that's okay!


I'm also adding a lot of nice nitpicky syntax comments. ;-) Enjoy. Hahaha


That being said, you did a great job on this project-- keep it up!

File: src/adagrams.js [hunk 1]
```diff
@@ -1,8 +1,113 @@
...
+      T: 6,
+      U: 4,
+      V: 2,
+      W: 2,
+      X: 1,
+      Y: 2,
+      Z: 1
+    }
+
+    // don't alter orginal array by creating new array and pushing letters into it
```
Comment (tildeee) [inline src/adagrams.js:32]: Your strategy here was to not alter the `pool` array within this method, so as to never alter it. And your strategy works! :+1: However, because of variable scope, you should feel free to alter it, because `const pool` will be re-created every time you call this method `drawLetters()`

File: src/adagrams.js [hunk 2]
```diff
@@ -1,8 +1,113 @@
...
+      const randomLetterIndex = Math.floor(Math.random() * (letters.length));
+      hand.push(letters[randomLetterIndex]);
+      letters.splice(randomLetterIndex, 1);
+    }
+    return hand;
+  },
+
+  usesAvailableLetters(input, lettersInHand) {
+    for (let i = 0; i < input.length; i += 1) {
+      let letter = input.charAt(i);
```
Comment (tildeee) [inline src/adagrams.js:52]: `letter` can be a `const`!

File: src/adagrams.js [hunk 3]
```diff
@@ -1,8 +1,113 @@
...
+      if (!lettersInHand.includes(letter)) {
+        return false;
+      }
+      let letterIndex = lettersInHand.indexOf(letter);
```
Comment (tildeee) [inline src/adagrams.js:56]: `letterIndex` can be a const!

File: src/adagrams.js [hunk 4]
```diff
@@ -1,8 +1,113 @@
...
+  scoreWord(word) {
+    const scoreKey = {
+      "A": 1, "B": 3, "C": 3, "D": 2,
+      "E": 1, "F": 4, "G": 2, "H": 4,
+      "I": 1, "J": 8, "K": 5, "L": 1,
+      "M": 3, "N": 1, "O": 1, "P": 3,
+      "Q": 10, "R": 1, "S": 1, "T": 1,
+      "U": 1, "V": 4, "W": 4, "X": 8,
+      "Y": 4, "Z": 10,
+    }
```
Comment (tildeee) [inline src/adagrams.js:72]: actually, JavaScript defaults these keys to strings, so you could do `const scoreKey = { A: 1, B: 3, C: 3}` etc

File: src/adagrams.js [hunk 5]
```diff
@@ -1,8 +1,113 @@
...
+
+    word = word.toUpperCase();
+    let letters = word.split('');
```
Comment (tildeee) [inline src/adagrams.js:75]: `letters` can be a `const`!

File: src/adagrams.js [hunk 6]
```diff
@@ -1,8 +1,113 @@
...
+        finalWinner.word = word;
+      } else if (this.scoreWord(word) === finalWinner.score) {
+        if (word.length === 10 && finalWinner.word.length  != 10) {
+          finalWinner.word = word;
+        } else if (word.length < finalWinner.word.length && (finalWinner.word.length != 10)) {
+          finalWinner.word = word;
+        }
+      }
+    }
+    return finalWinner
```
Comment (tildeee) [inline src/adagrams.js:106]: Despite the necessary nested code, this code style is visually easy to read. Well done!
